Checkpointing Feature

Lnaden opened the pull request:
This PR provides the ability for YANK to checkpoint simulations to save space and time when writing to file, at the cost of writing less frequently

This commit does the following:
* Splits all storage files into 2 files, one for checkpointing and one for analysis
* The two files are used for infrequent, and frequent writes respectively
* All file paths are now on a "base" and the individual files are derived from that base
* Most choices about which file gets which data is handled in the Reporter class
* Added `checkpoint_interval` option to specify how frequently writes occur on the checkpoint relative to the iterations
* Simulations are resumed only from the most recent checkpoint
* Data accessed from checkpoint file which is not on checkpoint_interval returns None objects
* Data cannot be written to checkpoint file at non-checkpoint_interval iterations (logger raises a warning, no action taken)
* Fixed an un-documented bug where the Health Reports were not being picked up correctly

andrrizzi commented on docs/yamlpages/options.rst:
```diff
@@ -497,6 +497,28 @@ Timestep of Langevin Dynamics production runs.
+you will get. If you want to write a checkpoint every iteration, set this to ``1``.
+
+Checkpoint information includes things like full coordinates and box vectors, as well as more static information such
+as metadata, simulation options, and serialized thermodynamic states.
```
This will need update if you end up moving stuff from the checkpoint storage.

andrrizzi commented on Yank/repex.py:
```diff
@@ -70,26 +70,50 @@ class Reporter(object):
-    def __init__(self, storage, open_mode=None):
-        self._storage_file_path = storage
-        self._storage = None
+    def __init__(self, base_storage, open_mode=None, checkpoint_interval=10):
```
I would make the constructor signature something like
```python
def __init__(self, storage, open_mode=None, checkpoint_interval=10, checkpoint_storage=None):
```
instead, and if `checkpoint_storage is None`, it will take the `storage` path and add a `_checkpoint` suffix to the name (plus append the correct extension). This way:
- We'd be able to pass a `Storage` objects to the constructor in the future in alternative to paths.
- We'd be able in the future to store positions and all the other informations in different formats by passing two different `Storage` objects or file paths with different extensions.

If a user wants to have a different name for the checkpoint than the default (probably a rare use case), he would pass directly a `Reporter` object instead of a file path. Something like:
```python
reporter = Reporter(storage='complex.nc', checkpoint_interval=50, checkpoint_storage='some_other_name.nc')
alchemical_phase.create(..., storage=reporter)
```

jchodera commented:
This sounds great!

How does resuming work? Does it back up to the last good checkpoint and *erase* the data from the non-checkpoint file? What would happen if, for example, `checkpoint_interval=100`, we fail at iteration 50 into the cycle, resume the simulation, and then terminate again before we reach iteration 100 in the cycle, and then try to analyze the data? Would we see a mix of old and new data?

Lnaden commented:
Whenever you resume, the file will always back up to the last good iteration it can, it then treats the rest of the data as garbage and overwrites it on the fly

Lnaden commented:
I'll double check that is what happens though

Lnaden commented:
Actually, the only way you would get junk data is to stop a simulation, resume it, stop it again before it makes it to the next checkpoint, then try to analyze that data

Lnaden commented:
Question: Do we want to allow the analysis file to move around (and be used for analysis) w/o the checkpoint file, or should they always be together?

The current implementation requires both files to be together. I can refactor it to allow the files to be separate for analysis (not resume),  but it will take some time.

jchodera commented:
It would be great if we could *analyze* the smaller analysis file without needing the checkpoint file.

andrrizzi commented:
I agree.

Lnaden commented:
Okay. I have re-refactored the code such that the analysis file can operate separate from the checkpoint file. Basic logic is:
```
Open analysis file
If write 
    Open new file for checkpoint
elif append or read 
    Try open checkpoint file
    Warn if no file
Functions acting on checkpoint file fail on their own
```

I still need to change the checkpoint behavior logic for resume, but what is here now will need looked over to make sure I didn't break something else

Looking over issues: Fixes #41

Lnaden commented:
Okay, I have added the ability for the reporter to ignore "junk" data, but still be properly indexed and sliced. I think this is probably ready for proper review again, @andrrizzi or @jchodera. 

What this does not do is allow the Reporter or Storage classes to be used as as part of the `create()` call. That might need to be done in a separate PR as this one is pretty feature rich already.

jchodera commented:
Awesome, thanks!

andrrizzi commented:
Whops! Somehow some of my comments got address to a line that changed. Sorry about that, they still refer to the last version of the code.

Lnaden commented on Yank/repex.py:
```diff
@@ -159,21 +251,57 @@ def open(self, mode='r'):
+        # Generate a unique UUID in case we need to assign it
+        # primary and subfiles are linked by UUID
+        # Uses uuid4 to avoid assigning hostname information
+        primary_uuid = str(uuid.uuid4())
```
I can't quite do that since the checkpoint file can move around and possibly be replaced. The UUID is really tied the primary file and so it can be safely written to as is. However, this takes into account the possibility of subfiles being left behind and the primary file being moved elsewhere post-creation. At that point, they would need the uuid from the existing file and not one created at run time, so I had to split the operation up into separate loops of "check the primary file" then "check the subfiles."

For the case of the checkpoint file, it could be reduced to that, but I'm thinking in case we use other subfiles later.

Lnaden commented on Yank/repex.py:
```diff
@@ -159,21 +251,57 @@ def open(self, mode='r'):
-                         "Using on file checkpoint interval of {}.".format(on_file_interval))
-            self._checkpoint_interval = on_file_interval
+
+        if self._storage_analysis is not None:
```
No, it could earlier when I had it check the checkpoint file, but I realized that we may need the checkpoint information for non-checkpoint file operations and changed the target file, I'll remove it

Lnaden commented on Yank/repex.py:
```diff
@@ -680,30 +806,29 @@ def write_timestamp(self, iteration):
+        if name in ['metadata']:  # store data needed for analysis in the primary file
+            storage = 'analysis'
+        else:
+            storage = 'checkpoint'
```
MCMC moves are not read by the analysis code
Acceptance rates are stored in their own variables (from `read_mixing_statistics`
```
n_accepted_matrix = self._storage_analysis.variables['accepted'][iteration, :, :].astype(np.int64)
        n_proposed_matrix = self._storage_analysis.variables['proposed'][iteration, :, :].astype(np.int64)```

Temperature is a problem. My big concern is that I don't like storing the massive serialized state strings in what is suppose to be the small analysis file, but if thats the only way we get Temperature, then I guess we'll have to

andrrizzi replied:
Sorry if I wasn't clear, what I mean is that many individual `MCMCMove` objects store their own `n_accepted` and `n_proposed` variables (e.g. `GHMCMove`, `MCRotationMove`) that I'm sure we'll want to read from the analysis soon in the future.

Lnaden commented:
I added the feature where `Repex` and all the `create()` methods can now take a `Reporter` class object as well as a string. I removed all the passing of the `checkpoint_interval` and `checkpoint_storage_file` from being passed between all the classes. YAMLBuild now also constructs the Reporter object and passes it to the file.

Added a method on the Reporter to get the primary file path so it can be used in strings and what not.

@andrrizzi could you take a quick look over this feature if you have time? The tests should all pass.

Lnaden commented:
I also think this should be the end of this PR and most of what we wanted to do with the Reporter class for now.

Advances the goals laid out in #535

Lnaden commented:
I also reduced the number of states of the "test `yank script...` command to speed up the test, its a major choke point of the tests and sometimes causes travis to time out. Of course, that may be due to the setup as well

andrrizzi commented:
Looks good!

Lnaden commented:
Woo!
